### PR TITLE
[CBRD-23280] Check loaddb authentication on client

### PR DESCRIPTION
**Pull request by radudoros**
http://jira.cubrid.org/browse/CBRD-23280

Make loaddb client retrieve class-name string from server, then check if we have insert authorization for the class-name.

### CODE REVIEW

**Comment by cubridci** [comment]
Build comment file: 
--------------
#### File: src/loaddb/load_db.c (1 incorrect indentation(s) found)
```
   212 	      catch (...)
```
--------------
**Please run 'indent -l120 -lc120' to correct the indentation**

--------------

File: src/loaddb/load_common.cpp [hunk 1]
```diff
@@ -537,7 +538,8 @@ namespace cubload
   }
 
   int
-  split (int batch_size, const std::string &object_file_name, class_handler &c_handler, batch_handler &b_handler)
+  split (int batch_size, const std::string &object_file_name, class_handler &c_handler, batch_handler &b_handler,
+	 auth_handler &a_handler)
```
Comment (ribram) [inline src/loaddb/load_common.cpp]: Can class_handler manage the check? It may be preferable to adding a new handler.

File: src/loaddb/load_db.c [hunk 2]
```diff
@@ -1237,15 +1237,38 @@ load_object_file (load_args * args)
 
       return ret;
     };
-  class_handler c_handler = [] (const batch &batch, bool &is_ignored) -> int
+  class_handler c_handler = [] (const batch &batch, bool &is_ignored, std::string &class_name) -> int
     {
-      int ret = loaddb_install_class (batch, is_ignored);
+      int ret = loaddb_install_class (batch, is_ignored, class_name);
```
Comment (ribram) [inline src/loaddb/load_db.c]: c_handler may check authorization.

File: src/loaddb/load_db.c [hunk 3]
```diff
@@ -1237,15 +1237,38 @@ load_object_file (load_args * args)
...
       delete &batch;
 
       return ret;
     };
+  auth_handler a_handler = [] (const std::string & class_name) -> int
+    {
+      DB_OBJECT *class_mop = db_find_class (class_name.c_str ());
```
Comment (ribram) [inline src/loaddb/load_db.c]: if current user is dba, checking authorization can be skipped.

File: src/loaddb/load_db.c [hunk 4]
```diff
@@ -1229,6 +1279,12 @@ load_object_file (load_args * args)
 	}
     }
 
+  if (!args->table_name.empty () && !load_has_authorization (args->table_name, AU_INSERT))
+    {
+      // user not authorized to insert in class
+      return ER_AU_NO_AUTHORIZATION;
```
Comment (ribram) [inline src/loaddb/load_db.c]: load_has_authorization may encounter other errors too (e.g. class does not exist). the right error should be propagated.

File: src/loaddb/load_db.c [hunk 5]
```diff
@@ -1194,6 +1199,44 @@ register_signal_handlers ()
   util_arm_signal_handlers (sig_handler, sig_handler);
 }
 
+static int
+load_has_authorization (const std::string & class_name, DB_AUTH au_type)
```
Comment (eseokoh) [inline src/loaddb/load_db.c]: Do you want `auth_type`? if the case, `DB_AUTH_INSERT` in line 1230 should be modified.

File: src/loaddb/load_db.c [hunk 6]
```diff
@@ -1202,6 +1245,12 @@ load_object_file (load_args * args)
   // first try to load directly on server
   if (!args->server_object_file.empty ())
     {
+      if (!au_is_dba_group_member (db_get_user ()))
+	{
+	  fprintf (stderr, "ERROR: Must be in dba group to load from a file on the server\n");
+	  error_code = ER_AU_NO_AUTHORIZATION;
```
Comment (eseokoh) [inline src/loaddb/load_db.c]: You may set ER_AU_DBA_ONLY error here.

File: src/loaddb/load_db.c [hunk 7]
```diff
@@ -1239,9 +1298,20 @@ load_object_file (load_args * args)
...
+      if (ret != NO_ERROR)
+      {
+	 return ret;
+      }
+
+      if (!is_ignored && !class_name.empty ())
+        {
+	  ret = load_has_authorization (class_name, AU_INSERT);
+        }
+
```
Comment (eseokoh) [inline src/loaddb/load_db.c]: I hope to have nicer indentation here.

**Pull request merged**